File: templates/typescript/src/utils/validation.ts
/**
 * Validation Utilities
 *
 * Input sanitization and validation helpers.
 * Security is critical — these protect against injection attacks,
 * path traversal, and other common vulnerabilities.
 */

import path from "node:path";
import { realpathSync } from "node:fs";

/**
 * Validates and normalizes a file path against allowed directories.
 * Prevents path traversal attacks (../../etc/passwd).
 *
 * Used by tools that access the filesystem.
 */
export function validatePath(
  requestedPath: string,
  allowedDirectories: string[]
): string {
  // Resolve to absolute path (handles .., symlinks, etc.)
  const resolved = path.resolve(requestedPath);

  // Resolve symlinks to get real path
  let realPath: string;
  try {
    realPath = realpathSync(resolved);
  } catch {
    // If file doesn't exist yet, use the resolved path
    realPath = resolved;
  }

  // Check if the real path falls within any allowed directory
  const isAllowed = allowedDirectories.some((dir) => {
    const resolvedDir = path.resolve(dir);
    return realPath.startsWith(resolvedDir + path.sep) || realPath === resolvedDir;
  });

  if (!isAllowed) {
    throw new Error(
      `Access denied: "${requestedPath}" is outside allowed directories`
    );
  }

  return realPath;
}

/**
 * Sanitizes a string to prevent command injection.
 * Strips shell metacharacters.
 */
export function sanitizeForShell(input: string): string {
  return input.replace(/[;&|`$(){}[\]<>!#*?~\n\r]/g, "");
}

/**
 * Validates a URL string.
 */
export function validateUrl(input: string): URL {
  try {
    return new URL(input);
  } catch {
    throw new Error(`Invalid URL: "${input}"`);
  }
}

File: templates/python/src/mcp_server_template/utils/validation.py
"""
Validation Utilities

Security helpers for input sanitization, path validation, etc.
"""

import os
from pathlib import Path


def validate_path(requested_path: str, allowed_directories: list[str]) -> Path:
    """
    Validate a file path against allowed directories.
    Prevents path traversal attacks.

    Args:
        requested_path: The path the user/LLM wants to access
        allowed_directories: List of directories that are allowed

    Returns:
        Resolved, validated Path

    Raises:
        ValueError: If path is outside allowed directories
    """
    resolved = Path(requested_path).resolve()

    for allowed_dir in allowed_directories:
        allowed = Path(allowed_dir).resolve()
        try:
            resolved.relative_to(allowed)
            return resolved
        except ValueError:
            continue

    raise ValueError(
        f"Access denied: '{requested_path}' is outside allowed directories"
    )


def require_env_vars(var_names: list[str]) -> dict[str, str]:
    """
    Validate that required environment variables are set.

    Args:
        var_names: List of required env var names

    Returns:
        Dict of var_name -> value

    Raises:
        EnvironmentError: If any variables are missing
    """
    values = {}
    missing = []

    for name in var_names:
        value = os.environ.get(name)
        if value:
            values[name] = value
        else:
            missing.append(name)

    if missing:
        raise EnvironmentError(
            f"Missing required environment variables: {', '.join(missing)}"
        )

    return values
